src/relay/pass/gradient.cc

```diff
@@ -172,6 +172,14 @@ struct ReverseAD : ExprFunctor<ADValue(const Expr &)> {
   }
 };
 
+/*! \brief Checks whether the annotation is defined.
+ *         If the annotation is defined, return it.
+ *         Otherwise, return a type hole.
+ */
+Type FromAnnotation(const Type& annotation) {
```

Comment:
I am not sure this is the best way to handle this issue, since it is essentially duplicating what type inference does. Having undefined types inside other types is rather disconcerting to me in general, but I am not sure how to dispel them in general. Could we make it so that a non-annotated type is made an incomplete type from the start? Would be curious as to your thoughts, @jroesch

Reply:
I think this is bug in Marisa's code, if he wants the system to do type inference for him it needs to insert incomplete types. If he wants the code to be polymorphic he needs to use polymorphism. It is invalid to have variables with `p->type_annotation` to null afaik.

Reply:
I think the issue is he is calling `Push` everywhere without providing a type, and expecting type inference, for example see line 134:

```
ll->Push(Add(args[i]->get<ADTensor>().reverse, rev[i]));
```

The push method will create a new variable with a null type annotation. I think if you change the call sites to provide an incomplete type

Reply:
I will investigate let lists, but I think the nulls are coming from function nodes and var nodes directly: A function node whose return type annotation is omitted has a null type for the ret_type field by default and a var node whose type annotation is omitted has a null type for the annotation field by default (I believe).